gemd/ingest/impl.py

```diff
@@ -0,0 +1,208 @@
+from gemd.entity.object.has_conditions import HasConditions
+from gemd.entity.object.has_parameters import HasParameters
+from gemd.entity.object.has_properties import HasProperties
+from gemd.entity.attribute import Property, PropertyAndConditions, Condition, Parameter
+from gemd.entity.template import PropertyTemplate, ParameterTemplate, ConditionTemplate, \
+    MaterialTemplate, MeasurementTemplate, ProcessTemplate
+from gemd.entity.template.attribute_template import AttributeTemplate
+from gemd.entity.bounds import RealBounds, IntegerBounds, CategoricalBounds, \
+    CompositionBounds, MolecularStructureBounds
+from gemd.entity.value import NominalReal, NominalInteger, NominalCategorical, \
+    EmpiricalFormula, Smiles
+from gemd.entity.value.base_value import BaseValue
+from typing import Union
+
+
+def make_node(name: str,
+              *,
+              process_name: str = None,
+              process_template: ProcessTemplate = None,
+              material_template: MaterialTemplate = None):
+    """
+    Generate a material-process spec-run quadruple.
+
+    Parameters
+    ----------
+    name: str
+        Name of the MaterialRun and MaterialSpec.
+    process_name: str
+        Name of the ProcessRun and ProcessSpec.  Defaults to
+        `process_template.name` if `process_template` is defined, else `name`.
+    process_template: ProcessTemplate
+        ProcessTemplate for the quadruple.
+    material_template: MaterialTemplate
+        MaterialTemplate for the quadruple.
+
+    Returns
+    --------
+    MaterialRun
+        A MaterialRun with linked processes, specs and templates
+
+    """
+    if process_name is None:
+        if process_template is None:
+            process_name = name
+        else:
+            process_name = process_template.name
+
+    my_process_spec = ProcessSpec(
+        name=process_name,
+        template=process_template
+    )
+
+    my_process_run = ProcessRun(
+        name=process_name,
+        spec=my_process_spec
+    )
+
+    my_mat_spec = MaterialSpec(
+        name=name,
+        process=my_process_spec,
+        template=material_template
+    )
+
+    my_mat_run = MaterialRun(
+        name=name,
+        process=my_process_run,
+        spec=my_mat_spec
+    )
+    return my_mat_run
+
+
+def add_edge(input_material: MaterialRun,
+             output_material: MaterialRun,
+             *,
+             name: str = None):
+    """
+    Connect two material-process spec-run quadruples with ingredients.
+
+    Parameters
+    ----------
+    input_material: MaterialRun
+        The `material` for the returned IngredientRun
+    output_material: MaterialRun
+        The `process` for the returned IngredientRun will be
+        `output_material.process`
+    name: str
+        The ingredient name.  Defaults to `input_material.name`.
+
+    Returns
+    --------
+    MaterialRun
+        A MaterialRun with linked processes, specs and templates
```

Comment:
It looks like it returns the ingredient spec

Reply:
It does.  Good catch on the copy-pasta.